File: Source/Entities/SolidcolorEntity.cs
using Celeste.Mod.Entities;
using Microsoft.Xna.Framework;
using Monocle;

namespace Celeste.Mod.CyrusSandbox.Entities
{
    [CustomEntity("CyrusHelper/SolidcolorEntity")]
    public class SolidcolorEntity : Entity
    {
        public Color color;
        public Color BorderColor;
        public enum InteractionType { None, Deadly, Solid };
        public readonly InteractionType interactions;
        public enum DrawType { Line, Fill, Bordered };
        public readonly DrawType drawType;
        public new int depth;
        private readonly Solid solid;
        private bool kill;
        public PlayerCollider pc;

        public SolidcolorEntity(EntityData data, Vector2 offset) : base(data.Position + offset)
        {
            kill = false;
            depth = data.Int("Depth", 8500);
            base.Depth = depth;
            color = data.HexColor("Color", Calc.HexToColor("#FFFFFF"));
            BorderColor = data.HexColor("LineColor", Calc.HexToColor("#000000"));
            drawType = data.Enum("DrawType", DrawType.Line);
            base.Collider = new Hitbox(data.Width, data.Height, 0f, 0f);
            interactions = data.Enum("Interaction", InteractionType.None);
            Add(pc = new PlayerCollider(OnCollide));
            switch (interactions)
            {
                case InteractionType.Deadly:
                    kill = true;
                    break;
                case InteractionType.Solid:
                    solid = new Solid(Position, base.Width, base.Height, safe: false);
                    break;
                case InteractionType.None:
                    break;
            }
        }

        private void OnCollide(Player player)
        {
            if (OnCollide != null && kill == true)
            {
                player.Die((player.Center - base.Center).SafeNormalize());
            }
        }

        public override void Added(Scene scene)
        {
            base.Added(scene);
            if (solid != null)
            {
                scene.Add(solid);
            }
        }

        public override void Render()
        {
            if (drawType == DrawType.Line)
            {
                Draw.HollowRect(Collider, color);
            }
            else if (drawType == DrawType.Fill)
            {
                Draw.Rect(Collider, color);
            }
            else if (drawType == DrawType.Bordered)
            {
                Draw.HollowRect(Collider, color);
                Draw.Rect(Collider, color);
            }
            base.Render();
        }

    }
}

File: Loenn/entities/SolidcolorEntity.lua
local utils = require("utils")
local drawable = require("structs.drawable_rectangle")

local SolidcolorEntity = {}

SolidcolorEntity.name = "CyrusSandbox/SolidcolorEntity"
SolidcolorEntity.minimumSize = {8,8}
SolidcolorEntity.fieldInformation = {
    Color = {
        fieldType = "color",
        allowXNAColors = true
    },
    LineColor = {
        fieldType = "color",
        allowXNAColors = true
    },
    Interaction = {
        options = {"None", "Deadly", "Solid"},
        editable = false
    },
    DrawType = {
        options = {"Line","Fill","Bordered"},
        editable = false
    }
}

SolidcolorEntity.placements = {
    {
        name = "Solid Color Entity",
        data = {
            width = 16,
            height = 16,
            Color = "ffffff",
            LineColor = "000000",
            DrawType = "Line",
            Interaction = "None",
            Depth = 8500,
            --color = "eebbdb" hehe eeby deeby
        }
    }
}

function SolidcolorEntity.sprite(room, entity)
    local x, y = entity.x or 0, entity.y or 0
    local width, height = entity.width or 8, entity.height or 8
    local color = entity.Color --custom
    local bordercolor = entity.LineColor
    local rect = drawable.fromRectangle(string.lower(entity.DrawType), x, y, width, height, color, bordercolor) --"line","fill","bordered"
    return rect
end

return SolidcolorEntity